vpr/src/route/rr_graph_storage.h

```diff
@@ -289,6 +289,11 @@ class t_rr_graph_storage {
         return edge_dest_node_[edge];
     }
 
+    // Get the source node for the specified edge.
+    RRNodeId edge_source_node(const RREdgeId& edge) const {
```

Comment:
Do not expose the edge_source_node outside of t_rr_graph_storage, it is not part of the public interface intentionally.

Reply:
This is needed for `create_edge_groups()': https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1289/files#diff-0ae086cbe871de8b37229e2748d822ceR3092

Is there a better way to access this data?

Reply:
The previous code at https://github.com/verilog-to-routing/vtr-verilog-to-routing/pull/1289/files#diff-0ae086cbe871de8b37229e2748d822ceL3149-L3157 didn't need direct access to RREdgeId -> source RRNodeId.

Reply:
Why can't there be an interface to iterate over all edges? It's possible to work around by first iterating nodes, then edges per node, but that could be less efficient if they are not in the same order.

What is the justification?

Reply:
RREdgeId -> source RRNodeId is only required early during graph build, and can likely be emptied out after partitioning is complete.  Having a public RREdgeId -> source RRNodeId precludes that modification.  Also, the relevant code doesn't random RREdgeId -> source RRNodeId access.  If you wanted to make a (source RRNodeId, RREdgeId) forward iterator, I'd be fine with that.

In general keeping N edge allocations where not needed is undesirable, and should be avoided unless absolutely required.

Reply:
Okay, understood. I'll add a method to iterate over all edges with source nodes to `t_rr_graph_storage`, which is what I really need here.

Reply:
I've added `for_each_edge` to map a function over all the edges in the most efficient order.